### PR TITLE
Support (re)connect time-out

### PR DESCRIPTION
**Pull request by iamqizhao**
based on #88. Thanks @gainward for the initial pull request.

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -50,30 +51,35 @@ var (
 	// ErrClientConnClosing indicates that the operation is illegal because
 	// the session is closing.
 	ErrClientConnClosing = errors.New("grpc: the client connection is closing")
+	// ErrClientConnTimeout indicates that the connection could not be
+	// established or re-established within the specified timeout.
+	ErrClientConnTimeout = errors.New("grpc: timed out trying to connect")
```
Comment (dsymonds) [inline clientconn.go]: would it be better to use context.DeadlineExceeded instead?

File: clientconn.go [hunk 2]
```diff
@@ -50,30 +51,35 @@ var (
...
 // credentials which will place auth state on each outbound RPC.
 func WithPerRPCCredentials(creds credentials.Credentials) DialOption {
-	return func(o *dialOptions) {
-		o.authOptions = append(o.authOptions, creds)
+	return func(o *transport.DialOptions) {
+		o.AuthOptions = append(o.AuthOptions, creds)
+	}
+}
+
+// WithTimeout returns a DialOption which configures a timeout for dialing a
```
Comment (dsymonds) [inline clientconn.go]: which -> that

File: clientconn.go [hunk 3]
```diff
@@ -50,30 +51,35 @@ var (
...
+// client connection.
```
Comment (dsymonds) [inline clientconn.go]: unwrap

File: clientconn.go [hunk 4]
```diff
@@ -133,12 +141,31 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 		if closeTransport {
 			t.Close()
 		}
-		newTransport, err := transport.NewClientTransport(cc.dopts.protocol, cc.target, cc.dopts.authOptions)
+		// Adjust timeout for the current try.
+		dopts := cc.dopts
+		if dopts.Timeout > 0 {
+			dopts.Timeout -= time.Since(start)
```
Comment (dsymonds) [inline clientconn.go]: I wouldn't change dopts.Timeout. It makes the logic too confusing. start.Add(dopts.Timeout) [when the timeout is set] should always be the cutoff.

File: credentials/credentials.go [hunk 5]
```diff
@@ -71,9 +71,15 @@ type Credentials interface {
...
-	Dial(addr string) (net.Conn, error)
+	// Dial connects to the given network address using net.Dial and then
+	// does the authentication handshake specified by the corresponding
+	// authentication protocol.
+	Dial(network, addr string) (net.Conn, error)
+	// DialWithDialer connects to the given network address using
+	// dialer.Dialand does the authentication handshake specified by the
+	// corresponding authentication protocol. Any timeout or deadline
+	// given in the dialer apply to connection and handshake as a whole.
+	DialWithDialer(dialer *net.Dialer, network, addr string) (net.Conn, error)
```
Comment (dsymonds) [inline credentials/credentials.go]: This doesn't seem like a good API. Perhaps move it to its own pull request to not distract from this one?

File: transport/transport.go [hunk 6]
```diff
@@ -310,10 +311,17 @@ func NewServerTransport(protocol string, conn net.Conn, maxStreams uint32) (Serv
 	return newHTTP2Server(conn, maxStreams)
 }
 
-// NewClientTransport establishes the transport with the required protocol
+// DialOptions covers all relevant options for dial a client connection.
```
Comment (dsymonds) [inline transport/transport.go]: "for dial a client connection" -> "for dialing a server"

File: transport/transport.go [hunk 7]
```diff
@@ -310,10 +311,17 @@ func NewServerTransport(protocol string, conn net.Conn, maxStreams uint32) (Serv
...
+	Protocol    string
+	AuthOptions []credentials.Credentials
+	Timeout     time.Duration
+}
+
+// NewClientTransport establishes the transport with the required DialOptions
 // and returns it to the caller.
-func NewClientTransport(protocol, target string, authOpts []credentials.Credentials) (ClientTransport, error) {
-	return newHTTP2Client(target, authOpts)
+func NewClientTransport(target string, opts DialOptions) (ClientTransport, error) {
```
Comment (dsymonds) [inline transport/transport.go]: this should take *DialOptions; the struct is of non-trivial size

File: clientconn.go (hunk 2 above)
Comment (iamqizhao) [inline clientconn.go]: done

File: clientconn.go (hunk 3 above)
Comment (iamqizhao) [inline clientconn.go]: done

File: clientconn.go (hunk 4 above)
Comment (iamqizhao) [inline clientconn.go]: dopts is just used to pass the latest timeout value to transport.NewClientTransport when for loop iterates (each iteration needs to have different timeout value for NewClientTransport when time goes forward). The cutoff is start.Add(cc.dopts.Timeout) always.

File: credentials/credentials.go (hunk 5 above)
Comment (iamqizhao) [inline credentials/credentials.go]: This is essential part of dial timeout support. It is used to pass timeout when TLS does dialing and handshake. This just mimics the API in the standard tls package (http://golang.org/pkg/crypto/tls/#DialWithDialer).

File: transport/transport.go (hunk 6 above)
Comment (iamqizhao) [inline transport/transport.go]: Done

File: transport/transport.go (hunk 7 above)
Comment (iamqizhao) [inline transport/transport.go]: done.

File: clientconn.go (hunk 1 above)
Comment (iamqizhao) [inline clientconn.go]: We do now have "context" concept here. So using it here is probably confusing.

**Comment by iamqizhao** [comment]
All comments are replied/addressed. Thanks.

**Comment by iamqizhao** [comment]
Seems there is still a loose end to tighten. Will ping this back once it is done.

File: clientconn.go [hunk 8]
```diff
@@ -133,12 +140,31 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
...
+		newTransport, err := transport.NewClientTransport(cc.target, &dopts)
 		if err != nil {
-			// TODO(zhaoq): Record the error with glog.V.
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+			sleepTime := backoff(retries)
+			// Fail early before falling into sleep.
+			if dopts.Timeout > 0 && dopts.Timeout < sleepTime + time.Since(start) {
```
Comment (gainward) [inline clientconn.go]: You've already removed time.Since(start) from dopts.Timeout, so I think it's sufficient to say dopts.Timeout > 0 && dopts.Timeout < sleepTime

File: credentials/credentials.go [hunk 9]
```diff
@@ -71,9 +71,15 @@ type Credentials interface {
...
 type TransportAuthenticator interface {
-	// Dial connects to the given network address and does the authentication
-	// handshake specified by the corresponding authentication protocol.
-	Dial(addr string) (net.Conn, error)
+	// Dial connects to the given network address using net.Dial and then
+	// does the authentication handshake specified by the corresponding
+	// authentication protocol.
+	Dial(network, addr string) (net.Conn, error)
+	// DialWithDialer connects to the given network address using
+	// dialer.Dialand does the authentication handshake specified by the
```
Comment (gainward) [inline credentials/credentials.go]: s/dialer.Dialand/dialer.Dial and/

File: transport/http2_client.go [hunk 10]
```diff
@@ -96,29 +96,38 @@ type http2Client struct {
...
 // and starts to receive messages on it. Non-nil error returns if construction
 // fails.
-func newHTTP2Client(addr string, authOpts []credentials.Credentials) (_ ClientTransport, err error) {
+func newHTTP2Client(addr string, opts *DialOptions) (_ ClientTransport, err error) {
 	var (
 		connErr error
 		conn    net.Conn
 	)
 	scheme := "http"
 	// TODO(zhaoq): Use DialTimeout instead.
```
Comment (gainward) [inline transport/http2_client.go]: remove todo

File: transport/http2_client.go [hunk 11]
```diff
@@ -96,29 +96,38 @@ type http2Client struct {
...
-	for _, c := range authOpts {
+	for _, c := range opts.AuthOptions {
 		if ccreds, ok := c.(credentials.TransportAuthenticator); ok {
 			scheme = "https"
 			// TODO(zhaoq): Now the first TransportAuthenticator is used if there are
 			// multiple ones provided. Revisit this if it is not appropriate. Probably
 			// place the ClientTransport construction into a separate function to make
 			// things clear.
-			conn, connErr = ccreds.Dial(addr)
+			if opts.Timeout > 0 {
```
Comment (gainward) [inline transport/http2_client.go]: If we limit opts.Timeout to >=0 elsewhere, this can be better expressed as:

conn, connErr = ccreds.DialWithDialer(&net.Dialer{Timeout: opts.Timeout}, "tcp", addr)

since ccreds.Dial would just use the same dialer by default, anyway.

File: clientconn.go (hunk 4 above)
Comment (gainward) [inline clientconn.go]: Can we bound dopts.Timeout >=0 here? That would simplify your dialing if/else statements.

File: transport/http2_client.go [hunk 12]
```diff
@@ -96,29 +96,33 @@ type http2Client struct {
...
 // and starts to receive messages on it. Non-nil error returns if construction
 // fails.
-func newHTTP2Client(addr string, authOpts []credentials.Credentials) (_ ClientTransport, err error) {
+func newHTTP2Client(addr string, opts *DialOptions) (_ ClientTransport, err error) {
 	var (
 		connErr error
 		conn    net.Conn
 	)
 	scheme := "http"
-	// TODO(zhaoq): Use DialTimeout instead.
```
Comment (iamqizhao) [inline transport/http2_client.go]: done.

File: clientconn.go (hunk 4 above)
Comment (iamqizhao) [inline clientconn.go]: done.

File: transport/http2_client.go [hunk 13]
```diff
@@ -96,29 +96,33 @@ type http2Client struct {
...
-	for _, c := range authOpts {
+	for _, c := range opts.AuthOptions {
 		if ccreds, ok := c.(credentials.TransportAuthenticator); ok {
 			scheme = "https"
 			// TODO(zhaoq): Now the first TransportAuthenticator is used if there are
 			// multiple ones provided. Revisit this if it is not appropriate. Probably
 			// place the ClientTransport construction into a separate function to make
 			// things clear.
-			conn, connErr = ccreds.Dial(addr)
+			dialer := &net.Dialer{Timeout: opts.Timeout}
```
Comment (iamqizhao) [inline transport/http2_client.go]: yes, we already have that guarantee. Done.

File: clientconn.go (hunk 8 above)
Comment (iamqizhao) [inline clientconn.go]: There was a bug here which is fixed. Your proposal is not working because  transport.NewClientTransport could take some time to return so that at this point the left time to finish dialing is cc.dopts.Timeout - time.Since(start).

**Comment by iamqizhao** [comment]
all addressed. Thanks.

File: clientconn.go [hunk 14]
```diff
@@ -133,12 +140,35 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
...
+		newTransport, err := transport.NewClientTransport(cc.target, &dopts)
 		if err != nil {
-			// TODO(zhaoq): Record the error with glog.V.
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+			sleepTime := backoff(retries)
+			// Fail early before falling into sleep.
+			if dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {
```
Comment (gainward) [inline clientconn.go]: Gotcha, this makes more sense. Maybe also check cc.dopts.Timeout >0 instead of dopts.Timeout. That's more of a stylistic choice, but since we're checking cc.dopts.Timeout in the right half of the expression, it makes sense.

File: transport/http2_client.go [hunk 15]
```diff
@@ -96,29 +96,32 @@ type http2Client struct {
...
 			// place the ClientTransport construction into a separate function to make
 			// things clear.
-			conn, connErr = ccreds.Dial(addr)
+			conn, connErr = ccreds.DialWithDialer(&net.Dialer{Timeout: opts.Timeout}, "tcp", addr)
 			break
 		}
 	}
 	if scheme == "http" {
-		conn, connErr = net.Dial("tcp", addr)
+		if opts.Timeout > 0 {
```
Comment (gainward) [inline transport/http2_client.go]: You can collapse this if/else into a single call to net.DialTimeout as well.

File: clientconn.go (hunk 14 above)
Comment (iamqizhao) [inline clientconn.go]: Actually we guarantee the total order between cc.dopts.Timeout and dopts.Timeout: cc.dopts.Timeout >= dopts.Timeout. So probably it is not needed?

File: transport/http2_client.go (hunk 15 above)
Comment (iamqizhao) [inline transport/http2_client.go]: done.

**Comment by iamqizhao** [comment]
All done. Thanks.

File: clientconn.go (hunk 14 above)
Comment (gainward) [inline clientconn.go]: Ah, I meant instead of. So

if cc.dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {

instead of

if dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {



But you're right that we're constrained cc.dopts.Timeout >= dopts.Timeout. I just thought it made more sense to use the latter var, since we're explicitly checking that it was set.

Comment (iamqizhao) [inline clientconn.go]: On Wed, Mar 4, 2015 at 2:30 PM, Wes <notifications@github.com> wrote:

> In clientconn.go
> <https://github.com/grpc/grpc-go/pull/93#discussion_r25821293>:
>
> >  		if err != nil {
> > -			// TODO(zhaoq): Record the error with glog.V.
> > +			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
> > +				cc.Close()
> > +				return ErrClientConnTimeout
> > +			}
> > +			sleepTime := backoff(retries)
> > +			// Fail early before falling into sleep.
> > +			if dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {
>
> Ah, I meant instead of. So
>
> if cc.dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime +
> time.Since(start) {
>
> instead of
>
> if dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {
>
> But you're right that we're constrained cc.dopts.Timeout >= dopts.Timeout.
> I just thought it made more sense to use the latter var, since we're
> explicitly checking that it was set.
>
make sense. Done.

File: clientconn.go [hunk 16]
```diff
@@ -133,12 +140,35 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 		if closeTransport {
 			t.Close()
 		}
-		newTransport, err := transport.NewClientTransport(cc.dopts.protocol, cc.target, cc.dopts.authOptions)
+		// Adjust timeout for the current try.
+		dopts := cc.dopts
```
Comment (dsymonds) [inline clientconn.go]: no need to copy the entire set of options. Only the timeout is relevant here. You can also collapse these two conditionals a bit.

```go
timeout := cc.dopts.Timeout
if timeout > 0 {
  timeout -= time.Since(start)
}
if timeout < 0 {
  ...
}
```

File: clientconn.go [hunk 17]
```diff
@@ -133,12 +140,35 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
...
+			return ErrClientConnTimeout
+		}
+		if dopts.Timeout > 0 {
+			dopts.Timeout -= time.Since(start)
+			if dopts.Timeout <= 0 {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+		}
+		newTransport, err := transport.NewClientTransport(cc.target, &dopts)
```
Comment (dsymonds) [inline clientconn.go]: oh, except this now wants a whole dialopts. hrm.

File: clientconn.go [hunk 18]
```diff
@@ -133,12 +140,35 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
...
 		if err != nil {
-			// TODO(zhaoq): Record the error with glog.V.
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
```
Comment (dsymonds) [inline clientconn.go]: I don't understand why this is here. If the operation timed out then the next time through this loop (or a few lines below) should notice that equally well.

Comment (dsymonds) [inline clientconn.go]: More particularly, a net.Error being a timeout does not guarantee that the full cc.dopts.Timeout has elapsed. It might be a sub-second failure due to name resolution, for instance.

Comment (iamqizhao) [inline clientconn.go]: make sense. Removed.

File: clientconn.go (hunk 17 above)
Comment (iamqizhao) [inline clientconn.go]: yep. Avoid bloating of input arguments.

**Comment by iamqizhao** [comment]
All done. Thanks.

**Comment by iamqizhao** [comment]
This blocks my other changes. Submit it now. Any post-submission comments are welcome.

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @iamqizhao isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_
